test/metabase/lib/equality_test.cljc

```diff
@@ -260,5 +262,30 @@
               {:name "DISCOUNT",   :selected? false}
               {:name "CREATED_AT", :selected? true}
               {:name "QUANTITY",   :selected? false}]
+             (mapv #(select-keys % [:name :selected?])
+                   (lib.equality/mark-selected-columns cols selected))
              (mapv #(select-keys % [:name :selected?])
                    (lib.equality/mark-selected-columns query cols selected)))))))
+
+(deftest ^:parallel index-of-closest-matching-metadata-test
+  (testing "index-of-closest-matching-metadata should find metadatas based on matching ID (#31482) (#33453)"
+    (let [query (lib/append-stage lib.tu/query-with-join)
+          cols  (lib/returned-columns query)
+          refs  (map lib.ref/ref cols)
+          a-ref [:field {:lib/uuid (str (random-uuid))} (meta/id :categories :name)]]
+      (is (=? [[:field {} "ID"]          ; 0
+               [:field {} "NAME"]        ; 1
+               [:field {} "CATEGORY_ID"] ; 2
+               [:field {} "LATITUDE"]    ; 3
+               [:field {} "LONGITUDE"]   ; 4
+               [:field {} "PRICE"]       ; 5
+               [:field {} "Cat__ID"]     ; 6
+               [:field {} "Cat__NAME"]]  ; 7
+              refs))
+      (testing "find-closest-matching-ref actually finds the wrong ref here! This is venues.name, not categories.name!!!"
+        (is (=? [:field {} "NAME"]
+                (lib.equality/find-closest-matching-ref query a-ref refs))))
+      (testing "... index-of-closest-matching-metadata finds the correct metadata, categories.name!!!"
+        (is (= 7
+               (lib.equality/index-of-closest-matching-metadata a-ref cols)
+               (lib.equality/index-of-closest-matching-metadata query a-ref cols)))))))
```

Comment:
Beautiful! This is 90% of the remaining issues with the viz-settings escalation branch solved and captured by a test. I was about to make these same changes today, splitting the matching to match refs against full columns, rather than refs.